src/sentry/integrations/gitlab/integration.py

```diff
@@ -0,0 +1,219 @@
+                "verify_ssl": form_data.get('verify_ssl')
+            })
+
+            return pipeline.next_step()
+
+        project_form = InstallationForm()
+
+        return render_to_response(
+            template='sentry/integrations/gitlab-config.html',
+            context={
+                'form': project_form,
+            },
+            request=request,
+        )
+
+
+class GitlabIntegrationProvider(IntegrationProvider):
+    key = 'gitlab'
+    name = 'Gitlab'
+    metadata = metadata
+    integration_cls = GitlabIntegration
+
+    needs_default_identity = True
+
+    features = frozenset([
+        IntegrationFeatures.ISSUE_BASIC,
+    ])
+
+    setup_dialog_config = {
+        'width': 1030,
+        'height': 1000,
+    }
+
+    def _make_identity_pipeline_view(self):
+        """
+        Make the nested identity provider view. It is important that this view is
+        not constructed until we reach this step and the
+        ``oauth_config_information`` is available in the pipeline state. This
+        method should be late bound into the pipeline vies.
+        """
+        identity_pipeline_config = dict(
+            oauth_scopes=(
+                'api',
+                'sudo',
+            ),
+            redirect_url=absolute_uri('/extensions/gitlab/setup/'),
+            **self.pipeline.fetch_state('oauth_config_information')
+        )
+
+        return NestedPipelineView(
+            bind_key='identity',
+            provider_key='gitlab',
+            pipeline_cls=IdentityProviderPipeline,
+            config=identity_pipeline_config,
+        )
+
+    def get_oauth_data(self, payload):
+        data = {'access_token': payload['access_token']}
+
+        # https://docs.gitlab.com/ee/api/oauth2.html#2-requesting-access-token
+        # doesn't seem to be correct, format we actually get:
+        # {
+        #   "access_token": "123432sfh29uhs29347",
+        #   "token_type": "bearer",
+        #   "refresh_token": "29f43sdfsk22fsj929",
+        #   "created_at": 1536798907,
+        #   "scope": "api sudo"
+        # }
+        if 'refresh_token' in payload:
+            data['refresh_token'] = payload['refresh_token']
+        if 'token_type' in payload:
+            data['token_type'] = payload['token_type']
+
+        return data
+
+    def get_group_info(self, access_token, installation_data):
+        session = http.build_session()
+        resp = session.get(
+            u'https://{}/api/v4/groups/{}'.format(
+                installation_data['url'], installation_data['group']),
+            headers={
+                'Accept': 'application/json',
+                'Authorization': 'Bearer %s' % access_token,
+            },
+            verify=installation_data['verify_ssl']
+        )
+
+        resp.raise_for_status()
+        return resp.json()
+
+    def get_pipeline_views(self):
+        return [InstallationConfigView(), lambda: self._make_identity_pipeline_view()]
```

Comment:
hmmm am I wrong to think `self._make_identity_pipeline_view` would work? Seems like `lambda` is unnecessary.